// @microsoft/power-apps/lib/internal/plugin/PluginMobile.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
let cordovaOperationsQueue = [];
function executePendingCordovaOperations() {
    for (let i = 0; i < cordovaOperationsQueue.length; i++) {
        cordovaOperationsQueue[i]();
    }
    cordovaOperationsQueue = [];
}
export function executePlugin(pluginName, pluginAction, params = []) {
    if (typeof cordova !== 'undefined') {
        cordova.exec(
        // eslint-disable-next-line no-empty-function
        () => { }, 
        // eslint-disable-next-line no-empty-function
        () => { }, pluginName, pluginAction, params);
    }
    else {
        cordovaOperationsQueue.push(() => {
            cordova.exec(
            // eslint-disable-next-line no-empty-function
            () => { }, 
            // eslint-disable-next-line no-empty-function
            () => { }, pluginName, pluginAction, params);
        });
    }
}
export function executePluginAsync(pluginName, pluginAction, params = []) {
    return new Promise((resolve, reject) => {
        if (typeof cordova !== 'undefined') {
            cordova.exec(resolve, reject, pluginName, pluginAction, params);
        }
        else {
            cordovaOperationsQueue.push(() => {
                cordova.exec(resolve, reject, pluginName, pluginAction, params);
            });
        }
    });
}
export async function initializeMobilePlugins() {
    await new Promise((resolve, reject) => {
        reject('Mobile is currently not supported');
        // if (document.readyState === 'loading') {
        //   document.addEventListener('DOMContentLoaded', () => {
        //     callLoader(resolve, reject);
        //   });
        // } else {
        //   callLoader(resolve, reject);
        // }
    });
    executePendingCordovaOperations();
}
export function isMobilePlayer() {
    const searchParams = new URLSearchParams(window.location.search);
    return searchParams.get('loaderType') === 'reactnative';
}
// async function getCordova(successCallback: () => void): Promise<void> {
//   if (typeof cordova !== 'undefined') {
//     successCallback();
//   } else {
//     document.addEventListener('deviceready', () => {
//       successCallback();
//     });
//   }
// }
// function callLoader(successCallback: () => void, failCallback: (reason: string) => void) {
// for ios, cordova script will be injected directly by player
// if (getSearchParam('injectCordova') === 'true' && getSearchParam('platformOS') === 'ios') {
//   getCordova(successCallback);
//   return;
// }
// const appServerUrl = getSearchParam('appServerUrl');
// const loaderPath = getSearchParam('loader');
// if (appServerUrl !== null && loaderPath !== null) {
//   addScript(
//     `${appServerUrl}${loaderPath.substring(1)}`,
//     () => {
//       getCordova(successCallback);
//     },
//     failCallback
//   );
// } else {
//    failCallback('Mobile is currently not supported');
// }
//
// function getSearchParam(param: string): string | null {
//   const searchParams = new URLSearchParams(window.location.search);
//   return searchParams.get(param);
// }
// function addScript(
//   src: string,
//   successCallback?: (event: Event) => void,
//   failCallback?: (event: ErrorEvent) => void
// ) {
//   const scriptTag = document.createElement('SCRIPT') as HTMLScriptElement;
//   scriptTag.setAttribute('crossorigin', 'anonymous');
//   scriptTag.type = 'text/javascript';
//   scriptTag.async = false;
//   if (successCallback) {
//     scriptTag.addEventListener('load', successCallback);
//   }
//   if (failCallback) {
//     scriptTag.addEventListener('error', failCallback);
//   }
//   document.head.appendChild(scriptTag);
//   scriptTag.src = src;
// }
