// @microsoft/power-apps/lib/__tests__/runtimeDataOperation.test.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { describe, expect, it } from '@jest/globals';
import { DefaultDataOperationOrchestrator } from '../internal/data/core/data/defaultOperationOrchestrator';
import { DataOperationErrorMessages } from '../internal/data/core/error/messages';
import { getMockConnectionsService } from './helpers/testHelpers';
import { convertOptionsToQueryString } from '../internal/data/core/data/executors/shared/stringQueryOptions';
describe('runtime data operation tests', () => {
    // This is the test class we can use to test private methods
    class TestRuntimeDataOperation extends DefaultDataOperationOrchestrator {
        validateOptions(options) {
            // @ts-expect-error: Accessing private method for testing
            return this._validateOptions(options);
        }
    }
    /*
      We test that this can take
      export interface IOperationOptions {
          maxPageSize?: number;
          select?: string[];
          filter?: string;
          orderBy?: string[];
          top?: number;
          skip?: number;
      }
  
      and covert it to standard OData params
      */
    let instance;
    // Create mock objects for the required dependencies
    /* eslint-disable @typescript-eslint/no-explicit-any */
    const mockDataverseOperation = {};
    const mockConnectorOperation = {};
    /* eslint-enable @typescript-eslint/no-explicit-any */
    beforeAll(() => {
        const dataSourceInfo = {
            tableId: 'tid',
            version: 'v1',
            apis: {
                op: {
                    parameters: [
                        {
                            name: 'userId',
                            in: 'header',
                            required: true,
                            type: 'string',
                        },
                        {
                            name: 'optionalHeader',
                            in: 'header',
                            required: false,
                            type: 'number',
                        },
                    ],
                    path: '/op',
                },
            },
        };
        const connectionsService = getMockConnectionsService(dataSourceInfo);
        instance = new TestRuntimeDataOperation(mockDataverseOperation, mockConnectorOperation, connectionsService);
    });
    describe('Convert options to string', () => {
        it('should convert select only', () => {
            const options = { select: ['field1', 'field2'] };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=field1%2Cfield2');
        });
        it('should convert filter only', () => {
            const options = { filter: 'name eq John' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$filter=name+eq+John');
        });
        it('should convert filter strings with leading or trailing spaces', () => {
            const options = { filter: '  name eq John  ' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$filter=name+eq+John');
        });
        it('should convert orderBy only to $orderby=name,age desc', () => {
            const options = { orderBy: ['name', 'age desc'] };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$orderby=name%2Cage%20desc');
        });
        it('should convert orderBy with mixed case to $orderby=Name,age DESC', () => {
            const options = { orderBy: ['Name', 'age DESC'] };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$orderby=Name%2Cage%20DESC');
        });
        it('should convert top only to $top=10', () => {
            const options = { top: 10 };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$top=10');
        });
        it('should convert skip only to $skip=5', () => {
            const options = { skip: 5 };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$skip=5');
        });
        it('should not convert maxPageSize to a URL param', () => {
            // maxPageSize is passed as a header, not a URL param
            const options = { maxPageSize: 100 };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('');
        });
        it('should convert count only to $count=true', () => {
            const options = { count: true };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$count=true');
        });
        it('should convert select and filter to $select=name&$filter=age gt 18', () => {
            const options = { select: ['name'], filter: 'age gt 18' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=name&$filter=age+gt+18');
        });
        it('should convert top and skip for pagination to $top=10&$skip=20', () => {
            const options = { top: 10, skip: 20 };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$top=10&$skip=20');
        });
        it('should convert orderBy and filter to $orderby=name&$filter=active eq true', () => {
            const options = { orderBy: ['name'], filter: 'active eq true' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$filter=active+eq+true&$orderby=name');
        });
        it('should convert all parameters to all OData parameters combined with &', () => {
            const options = {
                select: ['name', 'age'],
                filter: 'active eq true',
                orderBy: ['name desc'],
                top: 10,
                skip: 5,
                maxPageSize: 100,
                count: true,
            };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=name%2Cage&$filter=active+eq+true&$orderby=name%20desc&$top=10&$skip=5&$count=true');
        });
        it('should convert empty options to an empty string', () => {
            const options = {};
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('');
        });
        it('should convert empty arrays for select and orderBy to an empty string', () => {
            const options = { select: [], orderBy: [] };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('');
        });
        it('should convert null or undefined values for select and filter to an empty string', () => {
            const options = { select: null, filter: undefined };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('');
        });
        it('should convert single item arrays for select and orderBy to $select=id&$orderby=name', () => {
            const options = { select: ['id'], orderBy: ['name'] };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=id&$orderby=name');
        });
        it('should handle zero values for top, skip, and maxPageSize appropriately', () => {
            const options = { top: 0, skip: 0, maxPageSize: 0 };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$top=0&$skip=0');
        });
        it('should handle special characters in filter string as is', () => {
            const options = { filter: "name eq 'O'Reilly'" };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$filter=name+eq+%27O%27Reilly%27');
        });
        it('should ensure parameter names are correctly cased for OData', () => {
            const options = { select: ['Name', 'Age'], filter: 'Active eq true' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=Name%2CAge&$filter=Active+eq+true');
        });
        it('should convert orderBy with empty, whitespace, null, or undefined strings to $orderby=name', () => {
            const options = {
                orderBy: ['name', '', '  ', null, undefined],
            };
            expect(() => instance.validateOptions(options)).toThrow(new Error(`${DataOperationErrorMessages.InvalidOperationParameters}: orderBy must contain only non-empty strings`));
        });
        it('should throw if select contains empty, whitespace-only, null, or undefined strings', () => {
            const options = {
                select: ['id', '', '  ', null, undefined],
            };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if orderBy contains empty, whitespace-only, null, or undefined strings', () => {
            const options = {
                orderBy: ['name', '', '  ', null, undefined],
            };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if select is not an array', () => {
            const options = { select: 'id' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if orderBy is not an array', () => {
            const options = { orderBy: 'name' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if filter is not a string', () => {
            const options = { filter: 123 };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if top is not a number', () => {
            const options = { top: '10' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if skip is not a number', () => {
            const options = { skip: '5' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if maxPageSize is not a number', () => {
            const options = { maxPageSize: '100' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should throw if count is not a boolean', () => {
            const options = { count: 'true' };
            expect(() => instance.validateOptions(options)).toThrow();
        });
        it('should covert skip token parameter', () => {
            const options = { skipToken: '   abc123    ' };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$skiptoken=abc123');
        });
        it('should handle skip token with other parameters', () => {
            const options = {
                select: ['id'],
                filter: 'active eq true',
                skipToken: 'xyz789',
            };
            const result = convertOptionsToQueryString(options);
            expect(result).toBe('?$select=id&$filter=active+eq+true&$skiptoken=xyz789');
        });
    });
});
