// @microsoft/power-apps/lib/__tests__/mocks/mockDataverseRetrieveMultipleResponse.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { createMockArrayBuffer } from '../helpers/testHelpers';
export const mockDataverseResponse = {
    success: true,
    data: [
        {
            status: 200,
            headers: {
                allow: 'OPTIONS,GET,HEAD,POST',
                authactivityid: 'a79e2320-01dd-4178-a18c-db16575c6131',
                'cache-control': 'no-cache',
                'content-encoding': 'br',
                'Content-Type': 'application/json; odata.metadata=minimal; charset=utf-8',
            },
        },
        createMockArrayBuffer({
            '@odata.context': 'https://org394060f0.crm.dynamics.com/api/data/v9.0/$metadata#accounts',
            '@Microsoft.Dynamics.CRM.totalrecordcount': -1,
            '@Microsoft.Dynamics.CRM.totalrecordcountlimitexceeded': false,
            '@Microsoft.Dynamics.CRM.globalmetadataversion': '5652808',
            value: [
                {
                    accountnumber: '12345',
                    'participatesinworkflow@OData.Community.Display.V1.FormattedValue': 'No',
                    participatesinworkflow: false,
                    address2_country: null,
                },
            ],
            '@odata.nextLink': 'https://org394060f0.crm.dynamics.com/api/data/v9.0/accounts?$skiptoken=%3Ccookie%20pagenumber=%222%22%20pagingcookie=%22%253ccookie%2520page%253d%25221%2522%253e%253caccountid%2520last%253d%2522%257b717FCF9A-0CF6-EF11-BAE1-6045BD008D1B%257d%2522%2520first%253d%2522%257b84A47482-5252-F011-BEC1-000D3A3129E9%257d%2522%2520%252f%253e%253c%252fcookie%253e%22%20istracking=%22False%22%20/%3E',
        }),
    ],
};
