// @microsoft/power-apps/lib/__tests__/mocks/mockTeamsConnectorResponse.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import { createMockArrayBuffer } from '../helpers/testHelpers';
export const mockMSTeamsResponse = {
    success: true,
    data: [
        {
            status: 200,
            headers: {
                'client-request-id': 'bde8ffa9-e010-413a-bd9d-f0d96039f85b',
                'content-encoding': 'gzip',
                'Content-Type': 'application/json',
                Date: 'Wed, 16 Jul 2025 17:10:23 GMT',
                'odata-version': '4.0',
                'Request-Id': 'bde8ffa9-e010-413a-bd9d-f0d96039f85b',
                'strict-transport-security': 'max-age=31536000',
                'timing-allow-origin': '*',
                vary: 'Accept-Encoding',
                'x-ms-ags-diagnostic': '{"ServerInfo":{"DataCenter":"West US","Slice":"E","Ring":"4","ScaleUnit":"002","RoleInstance":"BY3PEPF00065EB2"}}',
                'x-ms-apihub-cached-response': 'true',
                'x-ms-apihub-obo': 'true',
                'x-ms-au-caller-id': 'e98928f0-2439-437d-9f0b-682b55a23053',
                'x-ms-au-creator-id': 'e98928f0-2439-437d-9f0b-682b55a23053',
                'x-ms-client-app-id': '/providers/Microsoft.PowerApps/apps/6e00c3f9-584a-44f7-a5a2-8d2c067ce2ed',
                'x-ms-client-environment-id': '/providers/Microsoft.PowerApps/environments/c9d64763-3dab-ed97-882b-f9347209b74c',
                'x-ms-client-object-id': 'e98928f0-2439-437d-9f0b-682b55a23053',
                'x-ms-client-region': 'tip1',
                'x-ms-client-session-id': '145ea95f-6a2c-40d5-acaf-ad0b8309712b',
                'x-ms-client-tenant-id': '91bee3d9-0c15-4f17-8624-c92bb8b36ead',
                'x-ms-dlp-ef': '-|-/-|-|-',
                'x-ms-dlp-gu': 'true|False',
                'x-ms-dlp-re': '-|-',
                'x-ms-environment-id': 'c9d64763-3dab-ed97-882b-f9347209b74c',
                'x-ms-flavor': 'Dogfood',
                'x-ms-mip-sl': '-|-|-|-',
                'x-ms-tenant-id': '91bee3d9-0c15-4f17-8624-c92bb8b36ead',
                'x-ms-user-agent': 'PowerApps/3.25072.2 (Web Player; AppName=6e00c3f9-584a-44f7-a5a2-8d2c067ce2ed)',
            },
            responseType: 'arraybuffer',
            duration: 505.0999999642372,
            responseSize: 4076,
            performanceInformation: {
                startTime: 58736090.80000001,
                name: 'https://c9d64763-3dab-ed97-882b-f9347209b74c.09.common.tip1002.azure-apihub.net/invoke',
                redirectStart: 0,
                redirectEnd: 0,
                fetchStart: 58736091.30000001,
                domainLookupStart: 58736091.30000001,
                domainLookupEnd: 58736091.30000001,
                connectStart: 58736091.30000001,
                connectEnd: 58736091.30000001,
                secureConnectionStart: 58736091.30000001,
                requestStart: 58736093.69999999,
                responseStart: 58736590.100000024,
                responseEnd: 58736591,
                transferSize: 1116,
                nextHopProtocol: 'h2',
            },
            sendHttpPipelinePerf: {
                sendHttpPipelinePerfTimestamps: {
                    SendHttpCore: {
                        pipelineStart: 1752685823250,
                        responseReceived: 1752685823754,
                        pipelineEnd: 1752685823754,
                    },
                    AppHttpClientPluginHandler: {
                        pipelineStart: 1752685823249,
                        responseReceived: 1752685823756,
                        pipelineEnd: 1752685823756,
                    },
                },
                transferSize: 1116,
                clientRequestId: 'fed77962-9d56-4057-8367-7809cf316139',
            },
        },
        // Convert JSON object to ArrayBuffer to simulate actual HTTP response
        createMockArrayBuffer({ randomKey: 'mockMSTeamsResponseData' }),
    ],
};
