// @microsoft/power-apps/lib/__tests__/mocks/mockLog.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-console */
export const createMockLogModule = (options = {}) => {
    const { silent = true } = options;
    return {
        Log: {
            createInstance: jest.fn((...args) => { }),
            resetInstance: jest.fn((...args) => {
                if (!silent) {
                    console.log('[MockLog.resetInstance]', ...args);
                }
            }),
            trackEvent: jest.fn((...args) => {
                if (!silent) {
                    console.log('â„¹ï¸ [MockLog.trackEvent]', ...args);
                }
            }),
            trackException: jest.fn((...args) => {
                if (!silent) {
                    console.log('ðŸ›‘ [MockLog.trackException]', ...args);
                }
            }),
            trackMetric: jest.fn((...args) => {
                if (!silent) {
                    console.log('ðŸ“Š [MockLog.trackMetric]', ...args);
                }
            }),
            startScenario: jest.fn((...args) => {
                if (!silent) {
                    console.log('ðŸŸ¢ [MockLog.startScenario]', ...args);
                }
            }),
            endScenario: jest.fn((...args) => {
                if (!silent) {
                    console.log('ðŸ [MockLog.endScenario]', ...args);
                }
            }),
            setDefaultProperties: jest.fn((...args) => {
                if (!silent) {
                    console.log('[MockLog.setDefaultProperties]', ...args);
                }
            }),
        },
    };
};
// Default export (silent)
export const mockSilentLogger = createMockLogModule({ silent: true });
/* eslint-enable @typescript-eslint/no-explicit-any */
/* eslint-enable no-console */
