// @microsoft/power-apps/lib/__tests__/mocks/silenceConsole.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
/**
 * Silences console output for the current test file
 * Call this in beforeEach or beforeAll
 */
export function silenceConsole() {
    const original = {
        log: console.log,
        error: console.error,
        warn: console.warn,
        info: console.info,
    };
    const mocks = {
        log: jest.fn(),
        error: jest.fn(),
        warn: jest.fn(),
        info: jest.fn(),
    };
    // Replace console methods with mocks
    console.log = mocks.log;
    console.error = mocks.error;
    console.warn = mocks.warn;
    console.info = mocks.info;
    const restore = () => {
        console.log = original.log;
        console.error = original.error;
        console.warn = original.warn;
        console.info = original.info;
    };
    return { mocks, restore };
}
/**
 * Higher-order function that silences console for specific methods only
 * example usage:
 * const { mocks, restore } = silenceSpecificConsole(['error', 'warn']);
 */
export function silenceSpecificConsole(methods) {
    const original = {};
    const mocks = {};
    methods.forEach((method) => {
        original[method] = console[method];
        mocks[method] = jest.fn();
        console[method] = mocks[method];
    });
    const restore = () => {
        methods.forEach((method) => {
            if (original[method]) {
                console[method] = original[method];
            }
        });
    };
    return { mocks, restore };
}
/* eslint-enable no-console */
/* eslint-enable @typescript-eslint/no-explicit-any */
