// @microsoft/power-apps/lib/__tests__/helpers/testHelpers.js
/*!
 * Copyright (C) Microsoft Corporation. All rights reserved.
 */
import path from 'path';
import process from 'process';
import { RuntimeDataSourceService } from '../../internal/data/core/metadata/runtimeDataSourceService';
// Get current directory for test files - use process.cwd() and resolve to this file's directory
export const testDir = path.resolve(process.cwd(), 'src', '__tests__');
// Test helpers for power-data-runtime tests
export function getMockDataClient(overrides = {}) {
    return {
        createDataAsync: jest.fn().mockResolvedValue({ success: true, data: { id: 1 } }),
        updateDataAsync: jest.fn().mockResolvedValue({ success: true, data: { id: 1 } }),
        deleteDataAsync: jest.fn().mockResolvedValue({ success: true }),
        retrieveDataAsync: jest.fn().mockResolvedValue({ success: true, data: { id: 1 } }),
        ...overrides,
    };
}
export function getMockMetadataClient(connectionReferences, overrides = {}) {
    return {
        getAppConnectionConfigsAsync: jest.fn().mockResolvedValue({ data: connectionReferences }),
        getAppDataSourceConfigsAsync: jest.fn().mockResolvedValue({ data: connectionReferences }),
        ...overrides,
    };
}
export function getMockClientProvider(dataClient, metadataClient, overrides = {}) {
    return {
        getDataClientAsync: jest.fn().mockResolvedValue(dataClient),
        getMetadataClientAsync: jest.fn().mockResolvedValue(metadataClient),
        ...overrides,
    };
}
export function getMockConnectionsService(dataSourceInfo, overrides = {}) {
    // Return a full instance of RuntimeDataSourceService with methods mocked, to match the type
    const service = Object.create(RuntimeDataSourceService.prototype);
    Object.assign(service, {
        getDataSource: jest.fn().mockResolvedValue(dataSourceInfo),
        initialize: jest.fn().mockResolvedValue(undefined),
        getUserDataSources: jest.fn().mockResolvedValue({}),
        hasDataSource: jest.fn().mockResolvedValue(true),
        ...overrides,
    });
    return service;
}
export function suppressConsoleErrorTelemetryLogger() {
    jest.spyOn(console, 'error').mockImplementation((...args) => {
        if (args[0] &&
            typeof args[0] === 'object' &&
            args[0].message &&
            args[0].message.includes('PowerDataRuntime.TelemetryLogger')) {
            return;
        }
    });
}
/**
 * Converts a JavaScript object to an ArrayBuffer for simulating HTTP response bodies in tests
 * This is needed because Jest does not support TextEncoder/Decoder in Node.js environment
 * @param data The object to convert
 * @returns ArrayBuffer containing the JSON string representation
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createMockArrayBuffer(data) {
    const jsonString = JSON.stringify(data);
    const buffer = Buffer.from(jsonString, 'utf8');
    return buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
}
